[FLINK-3849] [table] Add FilterableTableSource interface and rules for pushing it

KurtYoung opened the pull request:
This PR is based on #3166 , and added following changes:

1. Refactor `RexProgramExpressionExtractor` and `RexProgramExpressionExtractor` to `RexProgramExtractor` and `RexProgramRewriter`. `RexProgramExtractor` is responsible for retract either projection expressions or filter expression.
2. Make sure we don't fail during extracting and converting filter RexNodes to expressions. The expressions which successfully translated and unconverted RexNodes will both be returned.
3. Add some tests for `RexProgramExtractor`.
4. Provide unified `PushFilterIntoTableSourceScanRuleBase` to support filter push down in both batch and stream mode.
5. Add some logical tests for filter push down in different situations, like fully push down and partial push down.
6. Slight change of testing class `TestFilterableTableSource` to make it less specialized.

Another thing to notice is that UDF expression can not be translated from RexCall to Expression yet, since the name is not consistent with the name in FunctionCatalog due to the change of #3330. We can revisit this issue later since it's not very practicable for a TableSource can understand and execute UDF registered in Flink.

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/PushFilterIntoTableSourceScanRuleBase.scala:
```diff
@@ -0,0 +1,97 @@
+      filterableSource: FilterableTableSource,
+      description: String): Unit = {
+
+    if (filterableSource.isFilterPushedDown) {
```
tableSource should do not share between TableScan instances

KurtYoung commented:
To address the problem of not reusing `TableSource` when we create a new scan for table source, i changed some inheritance for current `BatchScan`, `StreamScan`, `BatchTableSourceScan`, `StreamTableSourceScan` and so on. The new structure is moe likely as the relationship between `FlinkTable` and `TableSourceTable`, `DataSetTable`, `DataStreamTable`. 

After changing the structure, it also make it possible to unify the push project into scan rule for both batch and stream mode.

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/RexProgramExtractor.scala:
```diff
@@ -0,0 +1,184 @@
+  */
+class InputRefVisitor extends RexVisitorImpl[Unit](true) {
+
+  private var fields = mutable.LinkedHashSet[Int]()
```
`var` → `val`

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/RexProgramExtractor.scala:
```diff
@@ -0,0 +1,184 @@
+
+    // return null if we cannot translate all the operands of the call
+    if (operands.contains(null)) {
+      Option.empty
```
I think, `None` is more brief than `Option.empty`

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -34,14 +34,24 @@ trait FilterableTableSource {
+    *         which contains all unsupported predicates.
     */
-  def applyPredicate(predicate: Array[Expression]): Array[Expression]
+  def applyPredicate(predicate: Array[Expression]): (FilterableTableSource[_], Array[Expression])
```
return `FilterableTableSource[T]` instead of `FilterableTableSource[_]`, otherwise in java, table source instance must be cast to `FilterableTableSource<?>`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/RexProgramExtractor.scala:
```diff
@@ -0,0 +1,184 @@
+  * @param inputNames      The input names of the relation node
+  * @param functionCatalog The function catalog
+  */
+class ConvertToExpression(
```
Rename to `RexNodeToExpressionConverter`?

KurtYoung replied:
Will change.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/RexProgramExtractor.scala:
```diff
@@ -0,0 +1,184 @@
+  }
+
+  override def visitLocalRef(localRef: RexLocalRef): Option[Expression] = {
+    throw new TableException("Bug: RefLocalRef should have been expanded")
```
`RefLocalRef` -> `RexLocalRef`

KurtYoung replied:
fixed

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/ProjectableTableSource.scala:
```diff
@@ -24,7 +24,7 @@ package org.apache.flink.table.sources
   * @tparam T The return type of the [[ProjectableTableSource]].
   */
-trait ProjectableTableSource[T] {
+trait ProjectableTableSource[T] extends TableSource[T] {
```
It's because i want to unify the PushProjectIntoScan rule codes for both batch and stream mode. And once we push down project into table source, we not only should create a new TableScan instance, but also a new TableSource instance. The codes are like:
```
val newTableSource = originTableSource.projectFields(usedFields)
// create a new scan with the new TableSource instance
val newScan = scan.copy(scan.getTraitSet, newTableSource) 
```
At first the `projectFields` method returned `ProjectableTableSource` which is not a `TableSource`, so i let `ProjectableTableSource` inherit from `TableSource`. But i just noticed we can just let `projectFields` return one `TableSource`, and problem resolved.

Will change this.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetCalc.scala:
```diff
@@ -42,63 +41,40 @@ class DataSetCalc(
+    calcProgram: RexProgram,
     ruleDescription: String)
-  extends SingleRel(cluster, traitSet, input)
+  extends Calc(cluster, traitSet, input, calcProgram)
```
This is because i want to unify the PushFilterIntoScan rule's code for both batch and stream mode. During executing the rule, we may need to create a new copy of the DataSetCalc or DataStreamCalc. It make things more easier to let these two classes inherit from `Calc`, and use `Calc.copy` to create a new copied instance. 

I do encountered some problem after i changed the hierarchy, some unit tests failed because of the plan changed. But it's because we don't calculate the cost for Calc right. I added some logic to `CommanCalc.computeSelfCost`, and everything works fine.

fhueske replied:
OK, sounds good to me. I was just curious, because I remembered we had some issues when `DataSetCalc` extended `Calc`. If everything works fine, I'm OK with this change.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/TableSource.scala:
```diff
@@ -39,4 +39,6 @@ trait TableSource[T] {
   def getReturnType: TypeInformation[T]
 
+  /** Describes the table source */
+  def explainTerms(pw: RelWriter): RelWriter = pw
```
Make sense to me, will change this.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/BatchTableSourceScan.scala:
```diff
@@ -58,16 +60,24 @@ class BatchTableSourceScan(
-    super.explainTerms(pw)
+    val terms = super.explainTerms(pw)
       .item("fields", TableEnvironment.getFieldNames(tableSource).mkString(", "))
+    tableSource.explainTerms(terms)
```
will change this.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/PushFilterIntoTableSourceScanRuleBase.scala:
```diff
@@ -0,0 +1,97 @@
+    // projection or filter exists.
+    val newScan = scan.copy(scan.getTraitSet, newTableSource)
+    val newRexProgram = {
+      if (remainingCondition != null || program.getProjectList.size() > 0) {
```
Thanks for the tips, will change this.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/schema/TableSourceTable.scala:
```diff
@@ -25,6 +25,7 @@ import org.apache.flink.table.sources.TableSource
 /** Table which defines an external table via a [[TableSource]] */
 class TableSourceTable[T](
     val tableSource: TableSource[T],
+    val tableEnv: TableEnvironment,
```
Yes, you are right, especially that UDF is currently registered as objects but not classes, it's really impossible to let TableSource supporting this. I will remove this filed and only use built-in functions when extracting expression form RexProgram.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/PushFilterIntoTableSourceScanRuleBase.scala:
```diff
@@ -0,0 +1,97 @@
+      filterableSource: FilterableTableSource[_],
+      description: String): Unit = {
+
+    if (filterableSource.isFilterPushedDown) {
```
Will change this.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/dataSet/PushFilterIntoBatchTableSourceScanRule.scala:
```diff
@@ -0,0 +1,55 @@
+    val scan: BatchTableSourceScan = call.rel(1).asInstanceOf[BatchTableSourceScan]
+    scan.tableSource match {
+      case _: FilterableTableSource[_] =>
+        calc.getProgram.getCondition != null
```
Will add

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/PushFilterIntoStreamTableSourceScanRule.scala:
```diff
@@ -0,0 +1,55 @@
+    val scan: StreamTableSourceScan = call.rel(1).asInstanceOf[StreamTableSourceScan]
+    scan.tableSource match {
+      case _: FilterableTableSource[_] =>
+        calc.getProgram.getCondition != null
```
Will add

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/RexProgramExtractor.scala:
```diff
@@ -0,0 +1,184 @@
+  override def visitOver(over: RexOver): Option[Expression] = Option.empty
+
+  private def lookupFunction(name: String, operands: Seq[Expression]): Option[Expression] = {
+    Try(functionCatalog.lookupFunction(name, operands)) match {
```
Yes, i've changed the passed in functionCatalog to an instance which only holds all built-in functions.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -0,0 +1,66 @@
+  * Adds support for filtering push-down to a [[TableSource]].
+  * A [[TableSource]] extending this interface is able to filter records before returning.
+  */
+trait FilterableTableSource[T] extends TableSource[T] {
```
Changed.

KurtYoung commented:
Hi @fhueske @godfreyhe, thanks for the review, i addressed most all your comments.

@fhueske Except for letting `TableSourceScan` be aware of whether filter has been pushed down. I'm not sure to let the `TableSourceScan` has this kind of information, i'd prefer to let them stay within the all kinds of `TableSource`. One drawback to let `TableSourceScan` has such kind of information is when we do the `TableSourceScan` copy, we need to take care all these information, make sure they also be copied correctly. In the future, if we add more extension to `TableSource` like we can push part of query down, we will face this problem. 

Regarding to the interface of `FilterableTableSource`, i agree with you that the trait containing some logic is not friendly with java extensions. So i removed the default implementation of `isFilterPushedDown`, the inherited class should take care of this method. And regarding the `Tuple2` thing, how about we pass in a mutable java list, and let table source to *pick out* expression from it and return a copy of table source which contains these pushed down predicates.

fhueske commented:
Thanks for the update @KurtYoung!

I think you have a good point about not extending `TableSourceScan`. The filter pushdown flag might not be the only one. Let's keep this information in `FilterableTableScan`. 

I'm not so sure about the interface of `applyPredicate(predicates: JList[Expression])`. I think we should not ask users to remove the accepted Expressions from the `predicates` list. How about we add a method `getPredicates(): JList[Expression]` which returns the accepted expressions. This method could also replace `isFilterPushedDown` if we expect `null` if `getPredicates()` was not called and an empty list if it was called but no predicate was accepted. IMO, `getPredicates()` has also the advantage that we can check that the `FilterableTableSource` did not modify the accepted predicates.

What do you think?

KurtYoung commented:
Hi @fhueske, you just pointed out a question i had when tonycox implementing the first version. Why we are preventing `FilterableTableSource` from modifying the expressions? I think it's totally up to them whether changing the expressions they took or keep them just they were. We have not reason to restrict the  `FilterableTableSource`'s behavior if they just do the things right.

Pass in a java list and told user who extending this to pick and remove the expressions the support is not super nice. But even if user just pick expressions but not remove them, we still get the correct answer.
 
What do you think?

fhueske commented:
Sure, a `FilterableTableSource` can implement its accepted predicates as it likes. 
However, I think we could expect that `getPredicates()` returns the accepted filter expressions as it got them. This makes it easier to check a correct behavior of the `FilterableTableSource`. Internally, the accepted expressions can be decomposed or otherwise changed.

Similarly, we would need to check that the Expressions which remain in the mutable list are also present in the input list.

fhueske commented:
OK, I thought about this. Let's stick to your mutable list approach. As you said it is safe, i.e., in worst case the table source does not remove expressions and we'll have a bit of overhead due to the unnecessary filters.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/util/RexProgramTestBase.scala:
```diff
@@ -16,106 +16,96 @@
-    rexProgram.getExprList.asScala.map(_.toString)
+  // select amount, amount * price as total
+  // where (amount * price < 100) AND (id > 6)
+  protected def buildSimpleRexProgram2(): RexProgram = {
```
How is this different from `buildSimpleRexProgram1()`?

KurtYoung replied:
Oh, my bad, they are the same, will remove one of them

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/util/RexProgramTestBase.scala:
```diff
@@ -16,106 +16,96 @@
-    rexBuilder = new RexBuilder(typeFactory)
-    allFieldTypes = List(VARCHAR, BIGINT, INTEGER, DOUBLE).map(typeFactory.createSqlType(_))
-  }
+abstract class RexProgramTestBase {
```
good idea to make this an abstract class

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/util/RexProgramExtractorTest.scala:
```diff
@@ -0,0 +1,347 @@
+  }
+
+  private def testExtractSinglePostfixCondition(
+      fieldIndex: Integer,
```
`fieldIndex` is not used

KurtYoung replied:
changed

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/util/RexProgramExtractorTest.scala:
```diff
@@ -0,0 +1,347 @@
+    val t2 = rexBuilder.makeExactLiteral(BigDecimal.valueOf(100L))
+
+    // unsupported now: amount.cast(BigInteger)
+    val cast = builder.addExpr(rexBuilder.makeCast(allFieldTypes.get(1), t0))
```
`cast` is not used

KurtYoung replied:
changed

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/TestFilterableTableSource.scala:
```diff
@@ -0,0 +1,142 @@
+
+  val fieldTypes: Array[TypeInformation[_]] = Array(STRING, LONG, INT, DOUBLE)
+
+  // all predicates with filed "amount"
```
filed -> field

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/TestFilterableTableSource.scala:
```diff
@@ -0,0 +1,142 @@
+      cnt <- 0 until recordNum
+      if shouldCreateRow(cnt)
+    } yield {
+      val row = new Row(fieldNames.length)
```
can be simplified to 
```
val row = Row.of(
        s"Record_$cnt",
        cnt.toLong.asInstanceOf[Object],
        cnt.toInt.asInstanceOf[Object],
        cnt.toDouble.asInstanceOf[Object])
```

KurtYoung replied:
Thanks for the tips!

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/PushFilterIntoTableSourceScanRuleBase.scala:
```diff
@@ -36,36 +43,34 @@ trait PushFilterIntoTableSourceScanRuleBase {
+    val remainingPredicates = new util.LinkedList[Expression]()
+    predicates.foreach(e => remainingPredicates.add(e))
+
+    val newTableSource = filterableSource.applyPredicate(remainingPredicates)
```
I think we don't have to restrict this. If the user for some reason indeed want to change the predicates which returns back and  executed by framework, we should allow them to do so.

fhueske replied:
That was actually one of the reasons for converting the predicate to CNF. In CNF you can either evaluate a complete term or skip it and let a Calc evaluate it later. I don't see a case where it would make sense to modify a conjunctive term but not evaluating it. Of course, the table source could reorganize the expressions but what would be the purpose of that? IMO, this could be error prone and introduce bugs.

Do you have a concrete use case in mind, when a `FilterableTableSource` would modify the expressions that it does not evaluate?

KurtYoung replied:
For example the condition looks like  "a OR b", if the table source knows that `a` will always be `false`, but not sure with `b`. So it can actually modify the condition to `b` to reduce framework cost

fhueske replied:
OK, you're right. Thanks for the example.

KurtYoung commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -20,47 +20,40 @@ package org.apache.flink.table.sources
 
 import org.apache.flink.table.expressions.Expression
 
+import scala.tools.nsc.interpreter.JList
```
sure

KurtYoung commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/TableSourceTest.scala:
```diff
@@ -0,0 +1,382 @@
+        .where("amount > 2 && func0(amount) < 32")
+
+    // wo don't fail during the optimization
+    tEnv.optimize(result.getRelNode)
```
Yes, i added the check with function's identifier

KurtYoung commented:
Hi @fhueske, thanks for the review. I addressed all your comments and will rebase to master to let travis check. Will merge this after build success.

fhueske commented:
Can you squash the commits of the PR into two commits before merging? One for @tonycox and one for your changes? Thanks!

KurtYoung commented:
Sure, so the result is there will be 2 commits to master repository?

fhueske commented:
yes
